### DISCUSSION TITLE
Path towards an asynchronous SearXNG

### DISCUSSION THREAD

**Discussion by mochaaP** [searxng/searxng#3136] [Ideas]
Hello fellow developers!

I've been thinking of making SearXNG's codebase to fit modern Python standards better, one of the keys towards this goal is transitioning to ASGI.

The current SearXNG architecture primarily relies on Flask, coupled with uWSGI as the application server. While this setup has served us well, the benefits of embracing ASGI are numerous, especially on scalability. Modern Python's asynchronous features can significantly improve performance for I/O-bound tasks on a single thread. This means we don't have to spin up many workers to serve a public instance (since a worker uses ~150MiB memory on its own, this will quickly escalate to a scalability issue).

The elephant in the room is that we could not simply do a swap and expect huge performance gains in a effortless manner. Migrating to ASGI unfortunately need us to replace Flask with an ASGI-native framework (e.g., Starlette), thus indicates a huge refactor.

So there could be two paths for us:

* a) Adapt the current codebase to a new framework
      + likely less effort
      + keep the compatibility on people's workflow
      - have to care about tech debts
* b) Rebuild the whole thing from the ground up
      + asynchronous in mind from day one
      + as a chance to restructure and clean up the codebase
      - needs much more effort

What are your thoughts? :)

**Comment by dalf** [comment]
I've tried that a lot, and while the idea sounds good, there are too many obstacles. Only plan b) might work.

A few attempts:
* https://github.com/searx/searx/pull/1724
* https://github.com/searx/searx/pull/2313
* https://github.com/searxng/searxng/pull/256 (see https://github.com/encode/starlette/issues/279#issuecomment-505068010 )
* https://github.com/searx/searx/wiki/Milestones#milestone-12---async
* https://github.com/searx/searx/pull/1725 (more or less related to the topic)
* https://github.com/searxng/searxng/pull/254

There is a relic of these attempts in the current implementation: SearXNG uses the async API of httpx. See https://github.com/searxng/searxng/blob/master/searx/network/client.py

However we are on the path to remove all async code: https://github.com/searxng/searxng/pull/2685

Why it does not work IMHO:
* async is slow in Python; the HTTP/2 protocol is complex to the point where the CPU usage starts to be a problem with the image proxy (yes, with async code. Actually I think sync code might be faster). I've made numerous attempts to speed up the thing without success (for example, https://github.com/dalf/hpack_nghttp2 , and another attempts using Cython, mypyc, etc...)
* async is cooperative multitasking: if an engine is CPU intensive, everything will slow down. To avoid this, some shenanigan are required, like use a threadpool to call lxml for large document (I remember it improved the response time).

To sum up, I think that async Python does not fit the use case for SearXNG. An idea can be this: write the core components of SearXNG in Rust, and write the engines in Python (with pyo3 it should be "easy"). This is a very long road, but this is the opportunity to drop some of the legacies (see also https://github.com/searxng/searxng/issues/681 , https://github.com/searxng/searxng/issues/534 )

However, I might be wrong all along, 

---

Regarding memory footprint: https://github.com/searxng/searxng/discussions/2633

Side topic: 
* https://github.com/dalf/pyhttp-benchmark
* https://github.com/dalf/pyhttp-benchmark/blob/master/results/output.md (not up to date)

**Reply by return42** [reply]
Just my 5 cents ..

In my opinion, we should think about where the bottlenecks are and how we can improve the situation. The attempts to use an asynchronous model for the core have not yet shown the desired results.

> Modern Python's asynchronous features can significantly improve performance for I/O-bound tasks on a single thread. 

Aspect performance: the bottleneck are the response times when querying the origin engines  

> This means we don't have to spin up many workers to serve a public instance (since a worker uses ~150MiB memory on its own, this will quickly escalate to a scalability issue).

Aspect footprint: we have components in the core that we could outsource to separate services ... I can think of two topics off the top of my head (there will certainly be more).

- https://github.com/searxng/fasttext-predict
- https://github.com/searxng/searxng/blob/master/searx/data/engine_descriptions.json

Engine descriptions can be placed in the redis DB and for fasttext-predict a local service can be installed.

Aspect performance: catchphrase "lazy evaluation" .. by example "engine description" is needed only rare.

These are a few topics that come to my mind spontaneously ... without having done any deeper analysis ... but I think it would be important that we think in this somewhat differentiated direction, because I believe that we will not find the holy grail that will solve all our problems.

**Reply by mochaaP** [reply]
> async is slow in Python; the HTTP/2 protocol is complex to the point where the CPU usage starts to be a problem with the image proxy (yes, with async code. Actually I think sync code might be faster). I've made numerous attempts to speed up the thing without success (for example, [dalf/hpack_nghttp2](https://github.com/dalf/hpack_nghttp2) , and another attempts using Cython, mypyc, etc...)

Shall we offload the image proxy to Morty and ditch the built-in image_proxy entirely? In my opinion, Python is not the best fit for this scenario.

> async is cooperative multitasking: if an engine is CPU intensive, everything will slow down. To avoid this, some shenanigan are required, like use a threadpool to call lxml for large document (I remember it improved the response time).

lxml uses libxml & libxslt under the hood, which is a rather advanced XML toolkit. Since we are only doing queries, we could potentially switch to a faster XML parsing library like <https://github.com/zeux/pugixml> which is [3-5x faster](https://pugixml.org/benchmark.html) in general.

> Regarding memory footprint: #2633

Is there any prior profiling work on this? I'm relatively new to this codebase, so I would appreciate it if somebody could point me out, thanks!

**Reply by return42** [reply]
> lxml uses libxml & libxslt under the hood, which is a rather advanced XML toolkit. Since we are only doing queries, we could potentially switch to a faster XML parsing library like https://github.com/zeux/pugixml which is [3-5x faster](https://pugixml.org/benchmark.html) in general.

why not / but what's the gain? .. 

> the bottleneck are the response times when querying the origin engines

**Reply by dalf** [reply]
> Shall we offload the image proxy to Morty and ditch the built-in image_proxy entirely?

There were multiple issues with the setup of Morty, that's why a good enough solution was embedded.

> In my opinion, Python is not the best fit for this scenario.

Even if I guess the answer, may I ask why?

> Since we are only doing queries, we could potentially switch to a faster XML parsing library like https://github.com/zeux/pugixml which is [3-5x faster](https://pugixml.org/benchmark.html) in general.

I don't know pugixml, however HTML5 does not seems to be supported: https://github.com/zeux/pugixml/issues/116#issuecomment-248231266

> Is there any prior profiling work on this? I'm relatively new to this codebase, so I would appreciate it if somebody could point me out, thanks!

I've done some profiling with `tracemalloc` but I haven't published them. As far I remember, the 150Mo split into: 
* translations (60Mo)
* data from `searx.data` (actually, the huge datasets can be stored as SQLite files)
* other stuff

The relevant part of the code to check:
https://github.com/searxng/searxng/blob/047c7a7cf6d5ffc6df062ff4b9e6b71126dee063/searx/webapp.py#L1320-L1326
